homeassistant/components/recorder/migration.py

```diff
@@ -38,23 +38,31 @@ def migrate_schema(instance):
 
 def _apply_update(engine, new_version):
     """Perform operations to bring schema up to date."""
-    from sqlalchemy import Table
+    from sqlalchemy import Table, Index
     from . import models
 
-    if new_version == 1:
-        def create_index(table_name, column_name):
-            """Create an index for the specified table and column."""
-            table = Table(table_name, models.Base.metadata)
-            name = "_".join(("ix", table_name, column_name))
-            # Look up the index object that was created from the models
-            index = next(idx for idx in table.indexes if idx.name == name)
-            _LOGGER.debug("Creating index for table %s column %s",
-                          table_name, column_name)
-            index.create(engine)
-            _LOGGER.debug("Index creation done for table %s column %s",
-                          table_name, column_name)
+    def create_index(table_name, column_name):
+        """Create an index for the specified table and column."""
+        table = Table(table_name, models.Base.metadata)
+        name = "_".join(("ix", table_name, column_name))
+        # Look up the index object that was created from the models
+        index = next(idx for idx in table.indexes if idx.name == name)
+        _LOGGER.debug("Creating index for table %s column %s",
+                      table_name, column_name)
+        index.create(engine)
+        _LOGGER.debug("Index creation done for table %s column %s",
+                      table_name, column_name)
 
+    if new_version == 1:
         create_index("events", "time_fired")
+    elif new_version == 2:
+        # Create compound start/end index for recorder_runs
+        table = Table("recorder_runs", models.Base.metadata)
+        index = Index("ix_start_end", table.c.start, table.c.end)
+        index.create(engine)
+        # Create indexes for states
+        create_index("states", "last_updated")
+        create_index("states", "created")
```

Comment:
I managed to get a compound index in here for recorder_runs. It's not a huge table (at least in my own DB) relative to states, but was causing a non-indexed table-scan. I really wanted to convert the create_index function to handle single or multiple indexes, but given how the Index constructor works, I was having a heck of a time. In either case, not sure how pruned indexes (e.g. `ALTER TABLE ... ADD INDEX ix_string (string(8))`) works but I believe it's possible.

I mention that because some of the types within tables ideally should be an ENUM for performance, but from the standpoint of flexibility are varchars. A reasonable compromise for that is to use a small pruned index which saves space but still gives decent cardinality. That's outside the scope of this PR, but sort of explaining my thought process here.